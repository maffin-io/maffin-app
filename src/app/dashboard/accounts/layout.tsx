import React from 'react';

export const metadata = {
  title: 'Home',
};

export default function AccountsLayout({
  children,
}: React.PropsWithChildren): JSX.Element {
  return (
    <div>
      {children}
    </div>
  );
}
