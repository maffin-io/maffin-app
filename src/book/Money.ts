import * as currencies from '@dinero.js/currencies';
import type { Currency as DineroCurrency } from 'dinero.js';
import * as djs from 'dinero.js';

import { toAmountWithScale, moneyToString } from '@/helpers/number';

export default class Money {
  private _raw: djs.Dinero<number>;

  /**
   * A wrapper class for dinero object to reduce verbosity and overhead. The constructor
   * checks for how many decimals the amount has and calculates the scale automatically.
   *
   * @param amount - The amount to be considered in the money object. Can be a float
   * @param currency - The currency
   * @param scale - Number representing custom scale. Default is 0, if passed differently, that one
   * is taken into account
   */
  constructor(amount: number, currency: string, scale?: number) {
    const dineroCurrency = toCurrency(currency);

    if (scale === undefined) {
      const { amount: newAmount, scale: newScale } = toAmountWithScale(amount);
      this._raw = djs.dinero({
        currency: dineroCurrency,
        amount: newAmount,
        scale: newScale,
      });
    } else {
      this._raw = djs.dinero({
        currency: dineroCurrency,
        amount,
        scale,
      });
    }
  }

  get currency(): string {
    return djs.toSnapshot(this._raw).currency.code;
  }

  get raw(): djs.Dinero<number> {
    return this._raw;
  }

  /**
   * Represents the money object as a string with the currency standardizing
   * the scale to 2 as a default. If the scale is passed then it uses that.
   */
  toString(scale = 2): string {
    return djs.toDecimal(
      djs.transformScale(this._raw, scale),
      ({ value, currency }) => `${value} ${currency.code}`,
    );
  }

  /**
   * Represents the money object as a string with the currency standardizing
   * the scale to 2 as a default. If the scale is passed then it uses that.
   *
   * The result is a localized string with the currency as a symbol
   */
  format(scale = 2): string {
    return djs.toDecimal(
      djs.transformScale(this._raw, scale),
      ({ value, currency }) => moneyToString(Number(value), currency.code),
    );
  }

  toNumber(): number {
    return parseFloat(djs.toDecimal(this._raw));
  }

  isNegative(): boolean {
    return djs.isNegative(this._raw);
  }

  isZero(): boolean {
    return djs.isZero(this._raw);
  }

  /**
   * Adds b to the current object and returns a new one with the result
   *
   * @param b - Money to be added
   * @returns - Money with result
   */
  add(b: Money): Money {
    const result = djs.toSnapshot(djs.add(b.raw, this.raw));
    return new Money(result.amount, this.currency, result.scale);
  }

  /**
   * Subtracts b to the current object and returns a new one with the result
   *
   * @param b - Money to be subtracted
   * @returns - Money with result
   */
  subtract(b: Money): Money {
    const result = djs.toSnapshot(djs.subtract(this.raw, b.raw));
    return new Money(result.amount, this.currency, result.scale);
  }

  /**
   * Multiplies b to the current object and returns a new one with the result
   * Initially, we try to multiply without transforming b but sometimes dinerojs
   * returns that it is an invalid amount (I have no idea why). If it fails, then
   * we transform b to `amountWithScale` and multiply with that.
   *
   * @param b - amount to be multiplied to in scale 0
   * @returns - Money with result
   */
  multiply(b: number): Money {
    let result;

    try {
      result = djs.multiply(this.raw, b);
    } catch {
      result = djs.multiply(this.raw, toAmountWithScale(b));
    }

    const snapshot = djs.toSnapshot(result);
    return new Money(snapshot.amount, this.currency, snapshot.scale);
  }

  /**
   * Checks for equality with b by comparing amount and currency
   *
   * @returns - boolean
   */
  equals(b: Money): boolean {
    return djs.equal(this._raw, b._raw);
  }

  /**
   * Converts a Money object to the currency specified
   *
   * @param to - Money object containing the rate to
   */
  convert(
    to: string,
    rate: number,
  ): Money {
    const rates = {
      [to]: {
        ...toAmountWithScale(rate),
      },
    };
    const currency = toCurrency(to);

    try {
      // @ts-ignore
      const result = djs.toSnapshot(djs.convert(this.raw, currency, rates));
      return new Money(result.amount, currency.code, result.scale);
    } catch (e) {
      throw new Error(`Error converting ${this.format()} to ${to} with rates ${JSON.stringify(rates)}: ${e}`);
    }
  }
}

function toCurrency(currency: string): DineroCurrency<number> {
  // @ts-ignore
  let dineroCurrency = currencies[currency];
  if (!dineroCurrency) {
    dineroCurrency = {
      code: currency,
      base: 10,
      exponent: 2,
    };
  }
  return dineroCurrency;
}
