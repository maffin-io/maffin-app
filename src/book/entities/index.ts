export { default as Account } from './Account';
export { default as BankConfig } from './BankConfig';
export { default as Commodity } from './Commodity';
export { default as Price } from './Price';
export { default as Split } from './Split';
export { default as Transaction } from './Transaction';
export { default as Book } from './Book';
export { default as BaseEntity } from './BaseEntity';
