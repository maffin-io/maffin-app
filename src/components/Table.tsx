import React from 'react';
import {
  ColumnDef,
  TableOptions,
  useReactTable,
  flexRender,
  getCoreRowModel,
  getSortedRowModel,
  getExpandedRowModel,
  ColumnSort,
  ExpandedState,
  getPaginationRowModel,
} from '@tanstack/react-table';
import classNames from 'classnames';
import { FaSortDown, FaSortUp } from 'react-icons/fa';

import Pagination from '@/components/Pagination';

export type TableProps<T extends object> = {
  id: string,
  columns: ColumnDef<T>[],
  data: T[],
  initialSort?: ColumnSort,
  showHeader?: boolean,
  showPagination?: boolean,
  tdClassName?: string,
  isExpanded?: boolean,
} & Partial<TableOptions<T>>;

export default function Table<T extends object = {}>(
  {
    id,
    initialSort,
    showHeader = true,
    showPagination = false,
    tdClassName = 'px-6 py-4',
    isExpanded = false,
    ...props
  }: TableProps<T>,
): JSX.Element {
  const tableConfig: TableOptions<T> = {
    getCoreRowModel: getCoreRowModel(),
    getSortedRowModel: getSortedRowModel(),
    getExpandedRowModel: getExpandedRowModel(),
    enableExpanding: true,
    initialState: {
      ...props.initialState,
      sorting: (initialSort && [initialSort]) || undefined,
      expanded: isExpanded as ExpandedState,
    },
    ...props,
  };

  if (showPagination && !props.manualPagination) {
    tableConfig.getPaginationRowModel = getPaginationRowModel();
  }

  const table = useReactTable(tableConfig);

  return (
    <>
      <div className="relative overflow-hidden rounded-md">
        <table id={id} className="w-full text-sm text-left">
          {
            showHeader
            && (
              <thead className="bg-white dark:bg-dark-700">
                {table.getHeaderGroups().map((headerGroup) => (
                  <tr key={headerGroup.id}>
                    {headerGroup.headers.map(header => (
                      <th
                        key={header.id}
                        scope="col"
                        className="px-6 py-3"
                        onClick={header.column.getToggleSortingHandler()}
                      >
                        <div className="flex">
                          {flexRender(header.column.columnDef.header, header.getContext())}
                          {(
                            header.column.columnDef.header !== ''
                              && header.column.getCanSort()
                              && (
                                <div className="flex items-center cursor-pointer">
                                  <FaSortUp
                                    className={
                                      classNames(
                                        {
                                          'opacity-100': header.column.getIsSorted() === 'asc',
                                          'opacity-50': header.column.getIsSorted() !== 'asc',
                                        },
                                        'absolute',
                                      )
                                    }
                                  />
                                  <FaSortDown
                                    className={
                                      classNames(
                                        {
                                          'opacity-100': header.column.getIsSorted() === 'desc',
                                          'opacity-50': header.column.getIsSorted() !== 'desc',
                                        },
                                        'absolute',
                                      )
                                    }
                                  />
                                </div>
                              )
                          )}
                        </div>
                      </th>
                    ))}
                  </tr>
                ))}
              </thead>
            )
          }

          <tbody>
            {table.getRowModel().rows.map(row => (
              <tr key={row.id} className="border-b border-white/70 dark:border-dark-700">
                {row.getVisibleCells().map(cell => (
                  <td key={cell.id} className={tdClassName}>
                    {flexRender(cell.column.columnDef.cell, cell.getContext())}
                  </td>
                ))}
              </tr>
            ))}
          </tbody>
        </table>
      </div>

      {
        showPagination
        && <Pagination<T> table={table} />
      }
    </>
  );
}
