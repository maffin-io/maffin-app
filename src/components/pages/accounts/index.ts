export { default as AccountsTable } from './AccountsTable';
export { default as MonthlyTotalHistogram } from './MonthlyTotalHistogram';
export { default as NetWorthPie } from './NetWorthPie';
export { default as LatestTransactions } from './LatestTransactions';
