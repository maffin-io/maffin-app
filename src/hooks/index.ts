export { default as useDataSource, DataSourceContext } from './useDataSource';
export type { DataSourceContextType } from './useDataSource';
export { default as useBookStorage } from './useBookStorage';
export { default as useGapiClient } from './useGapiClient';
