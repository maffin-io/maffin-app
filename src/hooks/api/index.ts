import { DateTime } from 'luxon';
import { mutate, SWRResponse } from 'swr';
import useSWRImmutable from 'swr/immutable';

import { Commodity } from '@/book/entities';
import { InvestmentAccount } from '@/book/models';
import * as queries from '@/lib/queries';
import type { PriceDBMap } from '@/book/prices';
import type { AccountsMap } from '@/types/book';
import type { Account, Split, Transaction } from '@/book/entities';

export function useStartDate(): SWRResponse<DateTime> {
  const key = '/api/start-date';
  return useSWRImmutable(
    key,
    fetcher(queries.getEarliestDate, key),
  );
}

export function useMainCurrency(): SWRResponse<Commodity> {
  const key = '/api/main-currency';
  return useSWRImmutable(
    key,
    fetcher(queries.getMainCurrency, key),
  );
}

export function useCommodity(guid: string): SWRResponse<Commodity> {
  const key = `/api/commodities/${guid}`;

  const result = useSWRImmutable(
    key,
    fetcher(() => Commodity.findOneByOrFail({ guid }), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useCommodities(): SWRResponse<Commodity[]> {
  const key = '/api/commodities';

  const result = useSWRImmutable(
    key,
    fetcher(() => Commodity.find(), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  if (result.data) {
    result.data.forEach(
      (commodity: Commodity) => mutate(
        `/api/commodities/${commodity.guid}`,
        commodity,
        { revalidate: false },
      ),
    );
  }

  return result;
}

export function useLatestTxs(): SWRResponse<Transaction[]> {
  const key = '/api/txs/latest';
  return useSWRImmutable(
    key,
    fetcher(queries.getLatestTxs, key),
  );
}

export function useAccount(guid: string): SWRResponse<Account> {
  const key = `/api/accounts/${guid}`;

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getAccounts(guid), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useAccounts(): SWRResponse<AccountsMap> {
  const key = '/api/accounts';

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getAccounts(), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  if (result.data) {
    Object.values(result.data as AccountsMap).forEach(
      (account: Account) => mutate(`/api/accounts/${account.guid}`, account, { revalidate: false }),
    );
  }

  return result;
}

export function useAccountsMonthlyTotals(): SWRResponse<queries.MonthlyTotals> {
  const { data: accounts } = useAccounts();
  const { data: prices } = useSWRImmutable(
    '/api/prices',
    fetcher(() => queries.getPrices({}), '/api/prices'),
  );

  const key = '/api/monthly-totals';
  const result = useSWRImmutable(
    (accounts && prices) ? key : null,
    fetcher(
      () => queries.getMonthlyTotals(accounts as AccountsMap, prices),
      key,
    ),
  );

  return result;
}

export function useInvestment(guid: string): SWRResponse<InvestmentAccount> {
  const key = `/api/investments/${guid}`;

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getInvestment(guid), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useInvestments(): SWRResponse<InvestmentAccount[]> {
  const key = '/api/investments';

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getInvestments(), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  if (result.data) {
    result.data.forEach(
      (investment: InvestmentAccount) => mutate(
        `/api/investments/${investment.account.guid}`,
        investment,
        { revalidate: false },
      ),
    );
  }

  return result;
}

export function useSplits(guid: string): SWRResponse<Split[]> {
  const key = `/api/splits/${guid}`;
  return useSWRImmutable(
    key,
    fetcher(async () => queries.getSplits(guid), key),
  );
}

/**
 * Returns prices for a given commodity
 */
export function usePrices(c: Commodity | undefined): SWRResponse<PriceDBMap> {
  const key = `/api/prices/${c?.guid}`;
  return useSWRImmutable(
    c?.guid ? key : null,
    fetcher(async () => queries.getPrices({ from: c }), key),
  );
}

function fetcher(f: () => Promise<any>, key: string) {
  return async () => {
    const start = performance.now();
    const r = await f();
    const end = performance.now();
    console.log(`${key}: ${end - start}ms`);
    return r;
  };
}
