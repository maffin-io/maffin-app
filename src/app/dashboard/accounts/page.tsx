'use client';

import React from 'react';
import { DateTime } from 'luxon';

import AddAccountButton from '@/components/buttons/AddAccountButton';
import {
  AccountsTable,
  NetWorthPie,
  NetWorthHistogram,
  MonthlyTotalHistogram,
  LatestTransactions,
} from '@/components/pages/accounts';
import DateRangeInput from '@/components/DateRangeInput';
import Onboarding from '@/components/onboarding/Onboarding';
import * as API from '@/hooks/api';

export default function AccountsPage(): JSX.Element {
  const { data: earliestDate } = API.useStartDate();
  let { data: accounts } = API.useAccounts();
  const { isLoading } = API.useAccounts();

  const [selectedDate, setSelectedDate] = React.useState(DateTime.now());

  if (isLoading) {
    return (
      <div className="h-screen">
        <div className="flex text-sm h-3/4 place-content-center place-items-center">
          Loading...
        </div>
      </div>
    );
  }

  accounts = accounts || { root: { childrenIds: [] } };
  const showOnboarding = Object.keys(accounts).length === 1;

  return (
    <>
      <Onboarding show={showOnboarding} />
      <div className="flex items-center">
        <span className="text-xl font-medium">
          Your finances
        </span>
        <span className="ml-auto mr-3">
          <DateRangeInput
            earliestDate={earliestDate}
            dateRange={{
              start: selectedDate,
              end: selectedDate,
            }}
            onChange={(value: { start: DateTime, end: DateTime }) => setSelectedDate(value.start)}
            asSingle
          />
        </span>
        <div>
          <AddAccountButton />
        </div>
      </div>
      <div className="grid grid-cols-12 items-start items-top pb-4">
        <div className="grid grid-cols-12 col-span-3">
          <div className="col-span-12 p-4 mr-4 rounded-sm bg-gunmetal-700">
            <NetWorthPie
              selectedDate={selectedDate}
            />
          </div>
          <div className="col-span-12 p-4 mr-4 rounded-sm bg-gunmetal-700">
            <AccountsTable
              selectedDate={selectedDate}
              isExpanded={showOnboarding}
            />
          </div>
        </div>
        <div className="grid grid-cols-12 col-span-9">
          <div className="col-span-9 p-4 mb-4 h-100 mr-4 rounded-sm bg-gunmetal-700">
            <div className="flex h-full items-center">
              <NetWorthHistogram
                startDate={earliestDate}
                selectedDate={selectedDate}
              />
            </div>
          </div>
          <div className="col-span-3 p-4 mb-4 mr-4 rounded-sm bg-gunmetal-700">
            <LatestTransactions />
          </div>
          <div className="col-span-6 p-4 mr-4 rounded-sm bg-gunmetal-700">
            <MonthlyTotalHistogram
              accounts={accounts?.type_income?.childrenIds.map((guid: string) => accounts?.[guid])}
              title="Income"
              selectedDate={selectedDate}
            />
          </div>
          <div className="col-span-6 p-4 mr-4 rounded-sm bg-gunmetal-700">
            <MonthlyTotalHistogram
              accounts={accounts?.type_expense?.childrenIds.map((guid: string) => accounts?.[guid])}
              title="Expenses"
              selectedDate={selectedDate}
            />
          </div>
        </div>
      </div>
    </>
  );
}
