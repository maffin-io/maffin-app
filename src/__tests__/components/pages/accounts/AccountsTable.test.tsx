import React from 'react';
import { render, screen } from '@testing-library/react';
import type { SWRResponse } from 'swr';

import { AccountsTable } from '@/components/pages/accounts';
import Table from '@/components/Table';
import { Account } from '@/book/entities';
import Money from '@/book/Money';
import * as apiHook from '@/hooks/api';

jest.mock('@/components/Table', () => jest.fn(
  () => <div data-testid="Table" />,
));
const TableMock = Table as jest.MockedFunction<typeof Table>;

jest.mock('@/hooks/api', () => ({
  __esModule: true,
  ...jest.requireActual('@/hooks/api'),
}));

describe('AccountsTable', () => {
  beforeEach(() => {
    jest.spyOn(apiHook, 'useAccounts').mockReturnValue({ data: undefined } as SWRResponse);
    jest.spyOn(apiHook, 'useAccountsMonthlyTotals').mockReturnValue({ data: undefined } as SWRResponse);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('creates empty Table with expected params', async () => {
    const { container } = render(<AccountsTable />);

    await screen.findByTestId('Table');
    expect(Table).toHaveBeenLastCalledWith(
      {
        id: 'accounts-table',
        columns: [
          {
            header: '',
            id: 'name',
            enableSorting: false,
            accessorKey: 'name',
            cell: expect.any(Function),
          },
          {
            header: '',
            id: 'total',
            accessorFn: expect.any(Function),
            cell: expect.any(Function),
          },
        ],
        data: [],
        initialSort: {
          desc: true,
          id: 'total',
        },
        isExpanded: false,
        showHeader: false,
        showPagination: false,
        tdClassName: 'p-2',
        getSubRows: expect.any(Function),
      },
      {},
    );
    expect(container).toMatchSnapshot();
  });

  it('creates table with expected params', async () => {
    jest.spyOn(apiHook, 'useAccounts').mockReturnValue(
      {
        data: {
          root: {
            guid: 'root',
            name: 'Root',
            type: 'ROOT',
            childrenIds: ['a1', 'a2'],
          } as Account,
          a1: {
            guid: 'a1',
            name: 'Assets',
            description: 'description',
            commodity: {
              mnemonic: 'EUR',
            },
            type: 'ASSET',
            childrenIds: [] as string[],
            placeholder: true,
          },
          a2: {
            guid: 'a2',
            name: 'Salary',
            commodity: {
              mnemonic: 'EUR',
            },
            type: 'INCOME',
            childrenIds: [] as string[],
          },
        },
      } as SWRResponse,
    );
    jest.spyOn(apiHook, 'useAccountsMonthlyTotals').mockReturnValue(
      {
        data: {
          a1: {
            '01/2023': new Money(100, 'EUR'),
          },
          a2: {
            '01/2023': new Money(-100, 'EUR'),
          },
        },
      } as SWRResponse,
    );

    render(<AccountsTable />);

    await screen.findByTestId('Table');
    expect(Table).toBeCalledTimes(1);
    expect(Table).toHaveBeenLastCalledWith(
      {
        id: 'accounts-table',
        columns: [
          {
            header: '',
            id: 'name',
            enableSorting: false,
            accessorKey: 'name',
            cell: expect.any(Function),
          },
          {
            header: '',
            id: 'total',
            accessorFn: expect.any(Function),
            cell: expect.any(Function),
          },
        ],
        data: [
          {
            account: {
              guid: 'a1',
              name: 'Assets',
              type: 'ASSET',
              description: 'description',
              commodity: {
                mnemonic: 'EUR',
              },
              childrenIds: [],
              placeholder: true,
            },
            leaves: [],
            total: expect.any(Money),
          },
          {
            account: {
              guid: 'a2',
              name: 'Salary',
              type: 'INCOME',
              commodity: {
                mnemonic: 'EUR',
              },
              childrenIds: [],
            },
            leaves: [],
            total: expect.any(Money),
          },
        ],
        initialSort: {
          desc: true,
          id: 'total',
        },
        isExpanded: false,
        showHeader: false,
        showPagination: false,
        tdClassName: 'p-2',
        getSubRows: expect.any(Function),
      },
      {},
    );

    expect((Table as jest.Mock).mock.calls[0][0].data[0].total.toString()).toEqual('100.00 EUR');
    expect((Table as jest.Mock).mock.calls[0][0].data[1].total.toString()).toEqual('100.00 EUR');
  });

  it('renders Name column as expected when expandable and not expandded', async () => {
    render(<AccountsTable />);

    await screen.findByTestId('Table');
    expect(Table).toBeCalledTimes(1);
    const nameCol = TableMock.mock.calls[0][0].columns[0];

    expect(nameCol.cell).not.toBeUndefined();
    const { container } = render(
      // @ts-ignore
      nameCol.cell({
        row: {
          original: {
            account: {
              guid: 'assets',
              name: 'Assets',
              type: 'ASSET',
              childrenIds: ['a1'],
            } as Account,
            total: new Money(100, 'EUR'),
            leaves: [],
          },
          getCanExpand: () => true,
          getIsExpanded: () => false,
        },
      }),
    );

    await screen.findByText('Assets');
    expect(container).toMatchSnapshot();
  });

  it('renders Name column as expected when expandable and expanded', async () => {
    render(<AccountsTable />);

    await screen.findByTestId('Table');
    expect(Table).toBeCalledTimes(1);
    const nameCol = TableMock.mock.calls[0][0].columns[0];

    expect(nameCol.cell).not.toBeUndefined();
    const { container } = render(
      // @ts-ignore
      nameCol.cell({
        row: {
          original: {
            account: {
              guid: 'assets',
              name: 'Assets',
              type: 'ASSET',
              childrenIds: ['a1'],
            } as Account,
            total: new Money(100, 'EUR'),
            monthlyTotals: {},
            children: [],
          },
          getCanExpand: () => true,
          getIsExpanded: () => true,
        },
      }),
    );

    await screen.findByText('Assets');
    expect(container).toMatchSnapshot();
  });

  it('renders Name column as expected when not expandable', async () => {
    render(<AccountsTable />);

    await screen.findByTestId('Table');
    expect(Table).toBeCalledTimes(1);
    const nameCol = TableMock.mock.calls[0][0].columns[0];

    expect(nameCol.cell).not.toBeUndefined();
    const { container } = render(
      // @ts-ignore
      nameCol.cell({
        row: {
          original: {
            account: {
              guid: 'assets',
              name: 'Assets',
              type: 'ASSET',
              childrenIds: ['a1'],
              placeholder: true,
            } as Account,
            total: new Money(100, 'EUR'),
            monthlyTotals: {},
            children: [],
          },
          getCanExpand: () => false,
        },
      }),
    );

    await screen.findByText('Assets');
    expect(container).toMatchSnapshot();
  });

  it('renders Total column as expected', async () => {
    render(<AccountsTable />);

    await screen.findByTestId('Table');
    expect(Table).toBeCalledTimes(1);
    const totalCol = TableMock.mock.calls[0][0].columns[1];

    expect(
      // @ts-ignore
      totalCol.accessorFn({ total: new Money(10, 'EUR') }),
    ).toEqual(10);

    expect(totalCol.cell).not.toBeUndefined();
    const { container } = render(
      // @ts-ignore
      totalCol.cell({
        row: {
          original: {
            account: {
              guid: 'assets',
              name: 'Assets',
              type: 'ASSET',
              childrenIds: ['a1'],
            } as Account,
            total: new Money(10, 'EUR'),
            children: [],
          },
        },
      }),
    );

    await screen.findByText('â‚¬10.00');
    expect(container).toMatchSnapshot();
  });
});
