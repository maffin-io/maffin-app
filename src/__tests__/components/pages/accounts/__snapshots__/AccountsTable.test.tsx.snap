// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountsTable creates empty Table with expected params 1`] = `
<div>
  <div
    data-testid="Table"
  />
</div>
`;

exports[`AccountsTable renders Name column as expected when expandable and expanded 1`] = `
<div>
  <div
    class="flex pl-NaN"
  >
    <button
      type="button"
    >
      <svg
        class="mr-1 text-xs opacity-20"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M11.178 19.569a.998.998 0 0 0 1.644 0l9-13A.999.999 0 0 0 21 5H3a1.002 1.002 0 0 0-.822 1.569l9 13z"
        />
      </svg>
    </button>
    <a
      class="badge hover:text-slate-300 info"
      data-tooltip-id="assets"
      href="/dashboard/accounts/assets"
    >
      Assets
    </a>
  </div>
</div>
`;

exports[`AccountsTable renders Name column as expected when expandable and not expandded 1`] = `
<div>
  <div
    class="flex pl-NaN"
  >
    <button
      type="button"
    >
      <svg
        class="mr-1 text-xs opacity-20"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M5.536 21.886a1.004 1.004 0 0 0 1.033-.064l13-9a1 1 0 0 0 0-1.644l-13-9A1 1 0 0 0 5 3v18a1 1 0 0 0 .536.886z"
        />
      </svg>
    </button>
    <a
      class="badge hover:text-slate-300 info"
      data-tooltip-id="assets"
      href="/dashboard/accounts/assets"
    >
      Assets
    </a>
  </div>
</div>
`;

exports[`AccountsTable renders Name column as expected when not expandable 1`] = `
<div>
  <div
    class="flex pl-NaN"
  >
    <button
      class="cursor-default"
      type="button"
    >
      <svg
        class="mr-1 text-xs opacity-20"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 2C6.486 2 2 6.486 2 12c.001 5.515 4.487 10.001 10 10.001 5.514 0 10-4.486 10.001-10.001 0-5.514-4.486-10-10.001-10zm0 18.001c-4.41 0-7.999-3.589-8-8.001 0-4.411 3.589-8 8-8 4.412 0 8.001 3.589 8.001 8-.001 4.412-3.59 8.001-8.001 8.001z"
        />
      </svg>
    </button>
    <span
      class="badge cursor-default info"
      data-tooltip-id="assets"
    >
      Assets
    </span>
  </div>
</div>
`;

exports[`AccountsTable renders Total column as expected 1`] = `
<div>
  <span>
    â‚¬10.00
  </span>
</div>
`;
